### PR TITLE
distopia 0.4.0 compatibility changes

### PR DESCRIPTION
**Pull request by hmacdope**
Test implementation of distopia-0.3.0 distances API.

Changes made in this Pull Request:
-  Adds support for distopia 0.3.0 API

PR Checklist
------------
 - [X] Tests?
 - [X] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

## Developers certificate of origin
- [X] I certify that this contribution is covered by the LGPLv2.1+ license as defined in our [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE) and adheres to the [**Developer Certificate of Origin**](https://developercertificate.org/).

----
📚 Documentation preview 📚: https://mdanalysis--4734.org.readthedocs.build/en/4734/

### CODE REVIEW

**Comment by hmacdope** [comment]
@orbeckst @richardjgowers this is ready for a first look over. The failing tests are for a 180 degree dihedral where numpy returns np.pi and distopia returns -np.pi (equivalent in polar coordinates). Are we ok to change the test to account for this? There are also options for changing in distopia but at the cost of a lot of performance improvement.

**Comment by orbeckst** [comment]
I'd be ok with changing tests but adding a note to docs.

File: package/MDAnalysis/lib/_distopia.py [hunk 1]
```diff
@@ -36,17 +36,15 @@
 else:
     HAS_DISTOPIA = True
 
-from .c_distances import (
```
Comment (IAlibay) [inline package/MDAnalysis/lib/_distopia.py:55]: Please make sure you add a version check to the distopia import check - see chemfiles for an example.

File: package/MDAnalysis/lib/distances.py [hunk 2]
```diff
@@ -101,6 +109,8 @@
    :class:`~MDAnalysis.core.groups.AtomGroup` or an :class:`np.ndarray`
 .. versionchanged:: 2.5.0
    Interface to the `distopia`_ package added.
+.. versionchanged:: 2.7.0
+    Distopia support greatly expanded.
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: Indentation and more detail
```suggestion
   Distopia support greatly expanded (with distopia ≥ 0.3.0).
```

File: package/MDAnalysis/lib/_distopia.py [hunk 3]
```diff
@@ -36,17 +36,15 @@
...
+
 import warnings
 import numpy as np
 
 
 def calc_bond_distance_ortho(
     coords1, coords2: np.ndarray, box: np.ndarray, results: np.ndarray
 ) -> None:
-    distopia.calc_bonds_ortho_float(
+    distopia.calc_bonds_ortho(
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: Codecov says not covered by tests? Is codecov wrong? If it's right then need to make sure that the new code is actually called in the tests.

**Review by orbeckst — CHANGES_REQUESTED** [review]
I am not sure if the tests are actually run — if we can confirm that they run then I have no blockers.

ALso updated CHANGELOG, please.

**Comment by orbeckst** [comment]
FYI, upstream distopia 0.3.0 release is currently breaking all tests, see #4739 .

File: package/MDAnalysis/lib/distances.py [hunk 4]
```diff
@@ -61,14 +61,22 @@
 you wish to use is covered by distopia. For more information see the
 `distopia documentation`_.
 
-.. Table:: Functions available using the `distopia`_ backend.
-    :align: center
-
-    +-------------------------------------+-----------------------------------+
-    | Functions                           | Notes                             |
-    +=====================================+===================================+
-    | MDAnalysis.lib.distances.calc_bonds | Doesn't support triclinic boxes   |
```
Comment (IAlibay) [inline package/MDAnalysis/lib/distances.py:70]: What about the note below about triclinic boxes?

**Review by IAlibay — CHANGES_REQUESTED** [review]

File: package/MDAnalysis/lib/_distopia.py (hunk 3 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/_distopia.py]: Codecov is wrong, but I am not exactly sure why its not picking up the conditional.

File: package/MDAnalysis/lib/distances.py (hunk 4 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/distances.py:70]: Sorry, missed that, fixed.

**Comment by hmacdope** [comment]
@orbeckst  @IAlibay feature freeze aside this is probably ready for re-review.

File: package/MDAnalysis/lib/distances.py [hunk 5]
```diff
@@ -101,6 +105,8 @@
    :class:`~MDAnalysis.core.groups.AtomGroup` or an :class:`np.ndarray`
 .. versionchanged:: 2.5.0
    Interface to the `distopia`_ package added.
+.. versionchanged:: 2.7.0
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]:
```suggestion
.. versionchanged:: 2.8.0
```

File: package/MDAnalysis/lib/distances.py [hunk 6]
```diff
@@ -101,6 +105,8 @@
...
+   Distopia support greatly expanded (with distopia ≥ 0.3.0).
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: We do need 0.3.1 because this code cannot detected 0.3.0:
```suggestion
   Distopia support greatly expanded (with distopia ≥ 0.3.1).
```

File: package/CHANGELOG [hunk 7]
```diff
@@ -59,6 +59,8 @@ Fixes
  * Fix groups.py doctests using sphinx directives (Issue #3925, PR #4374)
 
 Enhancements
+ * Improve distopia backend support in line with new functionality available 
+   in `distopia 0.3.0+`
```
Comment (orbeckst) [inline package/CHANGELOG]: I know that the support is in 0.3.0 but we need 0.3.1 so let's be explicit
```suggestion
   in distopia >= 0.3.1 (PR #4734)
```
and include PR

File: testsuite/MDAnalysisTests/lib/test_distances.py [hunk 8]
```diff
@@ -829,47 +839,20 @@ def convert_position_dtype_if_ndarray(a, b, c, d, dtype):
...
-        'calc_self_distance_array_ortho',
-        'calc_self_distance_array_triclinic',
-    ])
-    with patch.dict("sys.modules", {"distopia": mock_distopia_030}):
-        with pytest.warns(RuntimeWarning,
-                          match="Install 'distopia>=0.2.0,<0.3.0' to"):
-            import MDAnalysis.lib._distopia
-        assert not MDAnalysis.lib._distopia.HAS_DISTOPIA
+    sys.modules.pop("MDAnalysis.lib._distopia", None)   
+    if HAS_DISTOPIA:
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/lib/test_distances.py]: For the test I'd ignore HAS_DISTOPIA and not add the conditional. In principle we _always_ want to test explicitly that the version check is working. By using the conditional you implicitly assume that HAS_DISTOPIA was set to False because of the version check.

File: testsuite/MDAnalysisTests/lib/test_distances.py [hunk 9]
```diff
@@ -829,47 +839,20 @@ def convert_position_dtype_if_ndarray(a, b, c, d, dtype):
...
+        with patch('distopia.__version__', '0.1.0'):
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/lib/test_distances.py]: This only works if we can actually import distopia so I now understand why you need the conditional.

I'd prefer the test to work regardless of distopia being present, which can be achieved with the previous approach of creating a mocked module.


### background on `patch()`

(I had to read up on `patch` to understand how it does its magic so in case anyone else is interested, here's the background.)

The docs for [patch()](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) say

> `unittest.mock.patch(target, new=DEFAULT, spec=None, create=False, spec_set=None, autospec=None, new_callable=None, **kwargs)`
> ...
> _target_ should be a string in the form 'package.module.ClassName'. The target is imported and the specified object replaced with the new object, **so the target must be importable from the environment you are calling [patch()](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch) from.**

(my emphasis). 

Just to confirm: patch needs an importable module if one relies on the default behavior:
```python
In [2]: from unittest.mock import patch

In [3]: import distopia
---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[3], line 1
----> 1 import distopia

ModuleNotFoundError: No module named 'distopia'

In [4]: with patch('distopia.__version__', '0.1.0'):
   ...:     import distopia
   ...:     print(distopa.__version__)
   ...:

---------------------------------------------------------------------------
ModuleNotFoundError                       Traceback (most recent call last)
Cell In[4], line 1
----> 1 with patch('distopia.__version__', '0.1.0'):
      2     import distopia
      3     print(distopa.__version__)
...
ModuleNotFoundError: No module named 'distopia'
```

File: package/MDAnalysis/lib/_distopia.py [hunk 10]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+    coords1: np.ndarray, coords2: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_distance_array_no_box(coords1, coords2, results=results)
+
+def calc_distance_array_ortho(
+    coords1: np.ndarray, coords2: np.ndarray, box: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_distance_array_ortho(coords1, coords2, box[:3], results=results)
+
+def calc_distance_array_triclinic(
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py:170]: Not tested.

Also the docs for `distance_array` do not indicate that distopia is available.

File: package/MDAnalysis/lib/_distopia.py [hunk 11]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+    coords1: np.ndarray, coords2: np.ndarray, box: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_distance_array_triclinic(coords1, coords2, box, results=results)
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: Not tested (says codecov)

Also the docs for `distance_array` do not indicate that distopia is available.

File: package/MDAnalysis/lib/_distopia.py (hunk 3 above)
Comment (orbeckst) [reply package/MDAnalysis/lib/_distopia.py]: Codecov now sees them as tested. ✅

File: package/MDAnalysis/lib/_distopia.py [hunk 12]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+
+def calc_self_distance_array(
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: Not tested + docs of self_distance_array

File: package/MDAnalysis/lib/_distopia.py [hunk 13]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+    coords: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_self_distance_array_no_box(coords, results=results)
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: Not tested + docs of self_distance_array

File: package/MDAnalysis/lib/_distopia.py [hunk 14]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+
+def calc_self_distance_array_ortho(
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py:183]: Not tested + docs of self_distance_array

File: package/MDAnalysis/lib/_distopia.py [hunk 15]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+    coords: np.ndarray, box: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_self_distance_array_ortho(coords, box[:3], results=results)
+
+def calc_self_distance_array_triclinic(
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py:189]: Not tested + docs of self_distance_array

File: package/MDAnalysis/lib/_distopia.py [hunk 16]
```diff
@@ -77,8 +81,69 @@ def calc_bond_distance(
...
+    coords1: np.ndarray, coords2: np.ndarray, coords3: np.ndarray, coords4: np.ndarray, box: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_dihedrals_triclinic(coords1, coords2, coords3, coords4, box, results=results)
+
+def calc_distance_array(
+    coords1: np.ndarray, coords2: np.ndarray, results: np.ndarray
+) -> None:
+    distopia.calc_distance_array_no_box(coords1, coords2, results=results)
+
+def calc_distance_array_ortho(
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: It's awesome that we can now use distopia for distance_array!! 🎉 

The docs for self_distance_array and distance_array need to be updated
- add distopia to backends
- add a versionchanged 2.8.0 that states that distopia is now available: it's likely a big performance boost so we should make clear which versions support this improvement

File: package/MDAnalysis/lib/distances.py [hunk 17]
```diff
@@ -61,29 +61,33 @@
...
+   +===============================================+
+   | MDAnalysis.lib.distances.calc_bonds           |
+   +-----------------------------------------------+
+   | MDAnalysis.lib.distances.calc_angles          |
+   +-----------------------------------------------+
+   | MDAnalysis.lib.distances.calc_dihedrals       |
+   +-----------------------------------------------+
+   | MDAnalysis.lib.distances.distance_array       | 
+   +-----------------------------------------------+
+   | MDAnalysis.lib.distances.self_distance_array  |
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py:79]: hooray – distance arrays!

Please also update the docs for the individual functions
- update available backends
- add a versionchanged

**Review by orbeckst — CHANGES_REQUESTED** [review]
Awesome to also have support for distance_array/self_distance_array!

I have a few comments on the docs (see inline) and minor suggestion for the test.

My major issue at the moment is that codecov does not mark the distance_array functions as tested. Given that it sees the bonds and angles, I wonder if we are really testing distopia-accelerated distance_array. Can you please double check? Thanks!

File: testsuite/MDAnalysisTests/lib/test_distances.py [hunk 18]
```diff
@@ -946,6 +939,19 @@ def test_bonds(self, box, backend, dtype, pos, request):
         assert_almost_equal(dists_pbc[3], 3.46410072, self.prec,
                             err_msg="PBC check #w with box")
 
+    @pytest.mark.parametrize("dtype", (np.float32, np.float64))
+    @pytest.mark.parametrize("backend", distopia_conditional_backend())
+    def test_results_inplace_all_backends(self,  backend, dtype,):
```
Comment (hmacdope) [inline testsuite/MDAnalysisTests/lib/test_distances.py]: Problematic test here.

File: package/MDAnalysis/lib/distances.py [hunk 19]
```diff
@@ -303,6 +312,14 @@ def distance_array(reference: Union[npt.NDArray, 'AtomGroup'],
     distances = _check_result_array(result, (refnum, confnum))
     if len(distances) == 0:
         return distances
+
```
Comment (hmacdope) [inline package/MDAnalysis/lib/distances.py]: @orbeckst @richardjgowers we have an issue which is that the distopia backend doesn't seem to fill the result array on return. ie post call for the other backends, `result == distances` outside the function while for distopia its passed through. Not sure if I am doing something simple wrong. 

 Offending example in a test on this PR.
 
 ```python
     @pytest.mark.parametrize("dtype", (np.float32, np.float64))
    @pytest.mark.parametrize("backend", distopia_conditional_backend())
    def test_results_inplace_all_backends(self,  backend, dtype,):
        N = 10
        c0 = np.ones(3 * N, dtype=dtype).reshape(N, 3) * 2
        c1 = np.ones(3 * N, dtype=dtype).reshape(N, 3) * 3

        result = np.zeros(N, dtype=np.float64)
        distances.calc_bonds(c0, c1, result=result, backend=backend)
        expected = np.ones(N, dtype=dtype) * 3**(1/2)
        # test the result array is updated in place
        assert_almost_equal(result, expected, self.prec, err_msg="calc_bonds inplace failed")
 ```

We can get around this with a copy, but shouldn't be nessecary right? The distopia core API works inplace. Im fairly sure I am just getting some basic scoping wrong.

Comment (hmacdope) [reply package/MDAnalysis/lib/distances.py]: This test shows distopia works in place. 

https://github.com/MDAnalysis/distopia/pull/179/files#diff-a0adc50f7cfca5f98b7c9829f3709cdfb652f5ed0325fffc37885bb1308b4f18R69

File: package/MDAnalysis/lib/_distopia.py [hunk 20]
```diff
@@ -28,57 +28,123 @@
...
 
 
-from .c_distances import (
-    calc_bond_distance_triclinic as _calc_bond_distance_triclinic_serial,
-)
 import numpy as np
 
 
 def calc_bond_distance_ortho(
     coords1, coords2: np.ndarray, box: np.ndarray, results: np.ndarray
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: Any good reason why `coords1` does not have a type?

File: package/MDAnalysis/lib/distances.py (hunk 19 above)
Comment (orbeckst) [reply package/MDAnalysis/lib/distances.py]: I'd look into any use of [@check_coords](https://github.com/hmacdope/mdanalysis/blob/4fafd51de84d5b89be0559a412acefde0040847c/package/MDAnalysis/lib/util.py#L1965) as it creates _copies_ by default.

Comment (orbeckst) [reply package/MDAnalysis/lib/distances.py]: (Maybe not relevant as you're concerned with results.)

File: package/MDAnalysis/lib/distances.py [hunk 21]
```diff
@@ -386,6 +410,14 @@ def self_distance_array(reference: Union[npt.NDArray, 'AtomGroup'],
     distances = _check_result_array(result, (distnum,))
     if len(distances) == 0:
         return distances
+
+    if backend == 'distopia':
+        # distopia requires that all the input arrays are the same type,
+        # while MDAnalysis allows for mixed types, this should be changed
+        # pre 0.3.0 release see issue #3707 
+        distances = distances.astype(np.float32)
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py:458]: This may create a copy.

File: package/MDAnalysis/lib/distances.py [hunk 22]
```diff
@@ -401,6 +433,11 @@ def self_distance_array(reference: Union[npt.NDArray, 'AtomGroup'],
              args=(reference, distances),
              backend=backend)
 
+    if backend == 'distopia':
+        # mda expects the result to be in float64, so we need to convert it back
+        # to float64, change for 3.0, see #3707
+        distances = distances.astype(np.float64)
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py:485]: This may create a copy.

Comment (orbeckst) [reply package/MDAnalysis/lib/distances.py:485]: You may have to instead put the numbers back into the original array to make in-place work.
```python
result[:] = distances

...

return result
```
or something of that kind.

**Review by orbeckst — COMMENTED** [review]
With the `astype` required for distopia you may be creating a copy so that the top-level in-place operation is not actually in-place anymore. ... I think.

File: package/MDAnalysis/lib/distances.py (hunk 19 above)
Comment (orbeckst) [reply package/MDAnalysis/lib/distances.py]: I think it's actually the `astype()` — see comment review.

**Comment by orbeckst** [comment]
@hmacdope did you look further into the copy vs in-place issue?

**Comment by hmacdope** [comment]
@orbeckst I'll try get back to this over holidays, sorry for delay.

**Comment by orbeckst** [comment]
Ok! I think getting this one working will be important for a MDA paper.

**Comment by orbeckst** [comment]
@hmacdope I resolved the merge conflicts but please double-check that I didn't miss anything.

File: package/MDAnalysis/lib/distances.py (hunk 22 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/distances.py:485]: I used this strategy

**Comment by hmacdope** [comment]
@orbeckst this is probably ready for another look.

File: testsuite/MDAnalysisTests/lib/test_distances.py [hunk 23]
```diff
@@ -920,9 +932,10 @@ def test_transforms(self, S_mol, tri_vec_box, box, backend):
             err_msg="Round trip 2 failed in transform",
         )
 
+    @pytest.mark.parametrize("backend", distopia_conditional_backend())
     def test_selfdist(self, S_mol, box, tri_vec_box, backend):
         S_mol1, S_mol2 = S_mol
-        R_coords = distances.transform_StoR(S_mol1, box, backend=backend)
+        R_coords = distances.transform_StoR(S_mol1, box, backend="serial")
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/lib/test_distances.py:940]: Why hard-code *serial* here instead of using `backend`?

Comment (orbeckst) [reply testsuite/MDAnalysisTests/lib/test_distances.py:940]: Should the comment below apply to the `R_coords` line?

Maybe add a short comment.

File: testsuite/MDAnalysisTests/lib/test_distances.py [hunk 24]
```diff
@@ -948,7 +961,7 @@ def test_selfdist(self, S_mol, box, tri_vec_box, backend):
         )
 
         # Do it again for input 2 (has wider separation in points)
-        R_coords = distances.transform_StoR(S_mol2, box, backend=backend)
+        R_coords = distances.transform_StoR(S_mol2, box, backend="serial")
```
Comment (orbeckst) [inline testsuite/MDAnalysisTests/lib/test_distances.py:968]: Same questions as above – explain why `backend="serial"`

**Review by orbeckst — CHANGES_REQUESTED** [review]
Quick first look (sorry for brevity), see minor comments and below:

- Could you please also make PEP8 happy?
- Coverage?

**Comment by orbeckst** [comment]
Something is not right with the macos-13 tests — they timed out at 55 mins and they are again taking much longer than all the other runners.

**Comment by IAlibay** [comment]
> Something is not right with the macos-13 tests — they timed out at 55 mins and they are again taking much longer than all the other runners.

This is unrelated to this PR as far as I can tell (we probably should re-open #4781?). It's been a while but iirc the macos13 images had a sudden performance drop at some point in the last year-ish and it never was resolved. We could drop the tests to be "non optional deps only" that might help keep things a bit sane for now.

File: package/MDAnalysis/lib/_distopia.py [hunk 25]
```diff
@@ -28,48 +28,53 @@
...
 else:
     HAS_DISTOPIA = True
 
-    # check for compatibility: currently needs to be >=0.2.0,<0.3.0 (issue
-    # #4740) No distopia.__version__ available so we have to do some probing.
-    needed_funcs = ["calc_bonds_no_box_float", "calc_bonds_ortho_float"]
-    has_distopia_020 = all([hasattr(distopia, func) for func in needed_funcs])
-    if not has_distopia_020:
+    # check for compatibility: currently needs to be >=0.3.1,
+    # some versions of `distopia` don't have a version attribute
```
Comment (IAlibay) [inline package/MDAnalysis/lib/_distopia.py:45]: Are these versions known? I.e. if we know that this only affects a certain set of releases and they are all <0.3.1, then can we skip the warning altogther and just assume that this only happens in old versions we don't care about?

(intent: I'm trying to completly bypass the follow-up "there are no tests for line 50" with a blunt solution)

File: package/MDAnalysis/lib/_distopia.py [hunk 26]
```diff
@@ -78,8 +83,71 @@ def calc_bond_distance_triclinic(
...
 ) -> None:
-    # redirect to serial backend
-    warnings.warn(
-        "distopia does not support triclinic boxes, using serial backend instead."
-    )
-    _calc_bond_distance_triclinic_serial(coords1, coords2, box, results)
+    distopia.calc_bonds_triclinic(coords1, coords2, box, results=results)
+
+
+def calc_angle(
```
Comment (IAlibay) [inline package/MDAnalysis/lib/_distopia.py:87]: Has @RMeli lint `lib`? If so, could you run `black` on this file @hmacdope ?

File: package/MDAnalysis/lib/distances.py [hunk 27]
```diff
@@ -353,6 +370,13 @@ def distance_array(
             backend=backend,
         )
 
+    if backend == 'distopia':
```
Comment (IAlibay) [inline package/MDAnalysis/lib/distances.py]: Not for this PR, but you do this pattern enough that I wonder if life might be easier if you moved this into a decorator (or even just straight up into `check_coords`?

File: package/MDAnalysis/lib/distances.py [hunk 28]
```diff
@@ -1768,6 +1823,12 @@ def calc_angles(
                 backend=backend,
             )
 
+    if backend == 'distopia':
+        # mda expects the result to be in float64, so we need to convert it back
+        # to float64, change for 3.0, see #3707
+        angles = angles.astype(np.float64)
+        if result is not None:
+            result[:] = angles
```
Comment (IAlibay) [inline package/MDAnalysis/lib/distances.py:1831]: For here and dihedrals codecov is reporting no coverage, is that true/expected?

**Review by IAlibay — APPROVED** [review]
Couple things, apparent missing test coverage is the biggest thing, otherwise (assuming @orbeckst's comments are addressed), lgtm.

**Comment by orbeckst** [comment]
@hmacdope any bandwidth to get this one done soon-ish? It would be really good to have it in to finish one of the performance track items.

**Comment by IAlibay** [comment]
> @hmacdope any bandwidth to get this one done soon-ish? It would be really good to have it in to finish one of the performance track items.

If I can jump in here - it would be great to get this into a hypothetical v2.9.0 release of MDAnalysis (i.e. one that deals with #4899 and allows us to do py3.13).

**Comment by hmacdope** [comment]
@orbeckst @IAlibay yes can bump up the list, IIRC just need to add some coverage.

File: package/MDAnalysis/lib/distances.py (hunk 27 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/distances.py]: Will push to issue on PR completion

File: testsuite/MDAnalysisTests/lib/test_distances.py (hunk 23 above)
Comment (hmacdope) [reply testsuite/MDAnalysisTests/lib/test_distances.py:940]: Done, hardcode is because we template the test over the three backends for the function being tested (selfdist, [serial, openmp, distopia]) but were previously also using the same backend for `transform_StoR` as a helper function. This doesn't have distopia support so need to explicitly specify `serial`.

File: package/MDAnalysis/lib/distances.py (hunk 28 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/distances.py:1831]: Fixed now

**Comment by hmacdope** [comment]
@orbeckst this is probably ready for another round of review. I cut a new release of `distopia` to change some of the function names, should read much nicer now.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Sorry, just quick: there are distopia-related tests that are failing. Can you please check and fix or if necessary skip tests as needed?

File: package/MDAnalysis/lib/_distopia.py [hunk 29]
```diff
@@ -28,48 +28,52 @@
...
+    distopia_version = Version("0.0.0")
 else:
     HAS_DISTOPIA = True
 
-    # check for compatibility: currently needs to be >=0.2.0,<0.3.0 (issue
-    # #4740) No distopia.__version__ available so we have to do some probing.
-    needed_funcs = ["calc_bonds_no_box_float", "calc_bonds_ortho_float"]
-    has_distopia_020 = all([hasattr(distopia, func) for func in needed_funcs])
-    if not has_distopia_020:
+    # check for compatibility: currently needs to be >=0.3.1,
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: With the function renaming, needs to be >= 0.4.0, right?

File: package/MDAnalysis/lib/_distopia.py (hunk 25 above)
Comment (orbeckst) [reply package/MDAnalysis/lib/_distopia.py:45]: We can (now) just say that if we are not able to determine the version then it's too old.

File: package/MDAnalysis/lib/_distopia.py [hunk 30]
```diff
@@ -28,48 +28,52 @@
...
+    # some versions of `distopia` don't have a version attribute
+    try:
+        distopia_version = Version(distopia.__version__)
+    except AttributeError:
+        warnings.warn("distopia version cannot be determined, assuming 0.0.0")
```
Comment (orbeckst) [inline package/MDAnalysis/lib/_distopia.py]: Although we don't need to issue the warning, we need to be able to set the fake distopia version.

Is it worthwhile testing this code path with a proper mock? – @IAlibay ?

File: package/MDAnalysis/lib/distances.py [hunk 31]
```diff
@@ -101,6 +106,8 @@
    :class:`~MDAnalysis.core.groups.AtomGroup` or an :class:`np.ndarray`
 .. versionchanged:: 2.5.0
    Interface to the `distopia`_ package added.
+.. versionchanged:: 2.9.0
+   Distopia support greatly expanded (with distopia ≥ 0.3.1).
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: 0.4.0

File: package/MDAnalysis/lib/distances.py [hunk 32]
```diff
@@ -332,6 +341,14 @@ def distance_array(
     distances = _check_result_array(result, (refnum, confnum))
     if len(distances) == 0:
         return distances
+
+    if backend == "distopia":
+        # distopia requires that all the input arrays are the same type,
+        # while MDAnalysis allows for mixed types, this should be changed
+        # pre 0.3.0 release see issue #3707
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: Do you mean MDA 3.0.0 release (and not 0.3.0)?

File: package/MDAnalysis/lib/distances.py [hunk 33]
```diff
@@ -424,6 +450,14 @@ def self_distance_array(
     distances = _check_result_array(result, (distnum,))
     if len(distances) == 0:
         return distances
+
+    if backend == "distopia":
+        # distopia requires that all the input arrays are the same type,
+        # while MDAnalysis allows for mixed types, this should be changed
+        # pre 0.3.0 release see issue #3707
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: 3.0.0 instead of 0.3.0

File: package/MDAnalysis/lib/distances.py [hunk 34]
```diff
@@ -1630,13 +1671,17 @@ def calc_bonds(
     """
     numatom = coords1.shape[0]
     bondlengths = _check_result_array(result, (numatom,))
+    if backend == "distopia":
+        # distopia requires that all the input arrays are the same type,
+        # while MDAnalysis allows for mixed types, this should be changed
+        # pre 0.3.0 release see issue #3707
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: 3.0.0

File: package/MDAnalysis/lib/distances.py [hunk 35]
```diff
@@ -1746,6 +1794,13 @@ def calc_angles(
     numatom = coords1.shape[0]
     angles = _check_result_array(result, (numatom,))
 
+    if backend == "distopia":
+        # distopia requires that all the input arrays are the same type,
+        # while MDAnalysis allows for mixed types, this should be changed
+        # pre 0.3.0 release see issue #3707
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: 3.0.0

File: package/MDAnalysis/lib/distances.py [hunk 36]
```diff
@@ -1872,6 +1933,13 @@ def calc_dihedrals(
     numatom = coords1.shape[0]
     dihedrals = _check_result_array(result, (numatom,))
 
+    if backend == "distopia":
+        # distopia requires that all the input arrays are the same type,
+        # while MDAnalysis allows for mixed types, this should be changed
+        # pre 0.3.0 release see issue #3707
```
Comment (orbeckst) [inline package/MDAnalysis/lib/distances.py]: 3.0.0

**Review by orbeckst — CHANGES_REQUESTED** [review]
Please see small fixes in comments.

File: package/MDAnalysis/lib/_distopia.py (hunk 30 above)
Comment (IAlibay) [reply package/MDAnalysis/lib/_distopia.py]: Test is simple enough through monkeypatch. Adding it would be great, but I'm not sure it needs to be a blocker for this PR. At the end of the day we'll pin the conda-forge recipe to distopia>=0.4 and probably should do the same for pypi (is there a pypi release of distopia?).

TLDR; I'd be ok to see this is as a follow-up PR, but I agree we probably should do it at some point.

File: package/MDAnalysis/lib/distances.py (hunk 32 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/distances.py]: Yes fixed

File: package/MDAnalysis/lib/_distopia.py (hunk 30 above)
Comment (hmacdope) [reply package/MDAnalysis/lib/_distopia.py]: Can we push to seperate issue? Will raise when this is closed.

> (is there a pypi release of distopia?).

No there isn't, very hard to ship required libs.

**Review by orbeckst — APPROVED** [review]
lgtm, thanks !!!

excited to see it go in!

**Comment by hmacdope** [comment]
I will draft a blog post to announce
